components/resourceDetails/resourceSharing/addAgentButton/agentPickerModal/index.jsx

```diff
@@ -0,0 +1,376 @@
+
+export const handleSubmit = ({
+  newAgentsWebIds,
+  setNoAgentsAlert,
+  accessControl,
+  people,
+  addressBook,
+  mutatePermissions,
+  saveAgentToContacts,
+  onClose,
+  type,
+  fetch,
+}) => {
+  return () => {
+    if (!newAgentsWebIds.length) {
+      setNoAgentsAlert(true);
+      return;
+    }
+    newAgentsWebIds.forEach(async (agentWebId) => {
+      await accessControl.addAgentToNamedPolicy(agentWebId, type);
+      saveAgentToContacts(agentWebId, people, addressBook, fetch);
+      mutatePermissions();
+      onClose();
+    });
+  };
+};
+
+export const handleConfirmation = ({
+  open,
+  dialogId,
+  setConfirmationSetup,
+  setOpen,
+  setConfirmed,
+  handleSubmitNewWebIds,
+}) => {
+  return (confirmationSetup, confirmed) => {
+    if (open !== dialogId) return;
+    if (confirmationSetup && confirmed === null) return;
+    setConfirmationSetup(true);
+
+    if (confirmationSetup && confirmed) {
+      handleSubmitNewWebIds();
+    }
+
+    if (confirmationSetup && confirmed !== null) {
+      setConfirmed(null);
+      setOpen(null);
+      setConfirmationSetup(false);
+    }
+  };
+};
+
+export const handleSaveContact = async (iri, people, addressBook, fetch) => {
+  let response;
+  let error;
+
+  if (!iri) {
+    return;
+  }
+
+  try {
+    const { name, webId, types } = await fetchProfile(iri, fetch);
+    const existingContact = await findContactInAddressBook(
+      people,
+      webId,
+      fetch
+    );
+
+    if (existingContact.length) {
+      return;
+    }
+
+    if (name) {
+      const contact = { webId, fn: name };
+      response = await saveContact(addressBook, contact, types, fetch);
+    }
+  } catch (e) {
+    error = e; // setting the error in case we want to do something with it later (like displaying a notification to the user)
+  }
+  // eslint-disable-next-line consistent-return
+  return { response, error };
+};
+
+const useStyles = makeStyles((theme) => createStyles(styles(theme)));
+const VCARD_WEBID_PREDICATE = "https://www.w3.org/2006/vcard/ns#WebId";
+const TESTCAFE_ID_ADD_AGENT_PICKER_MODAL = "agent-picker-modal";
+const TESTCAFE_SUBMIT_WEBIDS_BUTTON = "submit-webids-button";
+
+export default function AgentPickerModal({
+  type,
+  text,
+  onClose,
+  mutatePermissions,
+  permissions,
+}) {
+  const [addressBook] = useAddressBook();
+  const {
+    data: people,
+    error: peopleError,
+    mutate: peopleMutate,
+  } = useContacts(addressBook, foaf.Person);
+
+  const classes = useStyles();
+  const { session } = useSession();
+  const { fetch } = session;
+  const { dataset } = useContext(DatasetContext);
+  const resourceIri = getSourceUrl(dataset);
+  const resourceName = getResourceName(resourceIri);
+  const [selectedTabValue, setSelectedTabValue] = useState("");
+  const { accessControl } = useContext(AccessControlContext);
+  const [newAgentsWebIds, setNewAgentsWebIds] = useState([]);
+  const [contactsArray, setContactsArray] = useState([]);
```

Comment:
I am using useContacts above to get the contacts dataset, but I still need to make an array of { thing, dataset } for the table.

Reply:
I am using useContacts a bit above that line, but I create the contacts array because I cannot use the dataset directly in the table. I need to create an array of { thing, dataset } for the table

Reply:
Just to note: With the new contact models that came in https://github.com/inrupt/pod-browser/commit/f87e5560f177d875790397e798a37618e3235004 the default form is `{ thing, dataset }`, so this will become easier.